src/consensus/pbft/pbft.h

```diff
@@ -372,11 +478,110 @@ namespace pbft
           break;
-        default:
-        {}
+        }
+        case pbft_append_entries:
+        {
+          AppendEntries r;
+
+          auto append_entries_index = store->current_version();
+
+          try
+          {
+            r =
+              channels->template recv_authenticated<AppendEntries>(data, size);
+          }
+          catch (const std::logic_error& err)
+          {
+            LOG_FAIL_FMT(err.what());
+            return;
+          }
+
+          LOG_TRACE_FMT(
+            "Append entries message from {}, my ae index is {}",
+            r.from_node,
+            append_entries_index);
+
+          auto node = nodes.find(r.from_node);
+          if (node != nodes.end())
+          {
+            node->second = r.idx;
+          }
+          else
+          {
+            nodes[r.from_node] = r.idx;
+          }
+
+          if (r.idx <= append_entries_index)
+          {
+            LOG_TRACE_FMT(
+              "Skipping append entries msg for index {} as we are at index {}",
+              r.idx,
+              append_entries_index);
+            break;
+          }
+
+          for (Index i = r.prev_idx + 1; i <= r.idx; i++)
+          {
+            append_entries_index = store->current_version();
+            LOG_TRACE_FMT("Recording entry for index {}", i);
+
+            if (i <= append_entries_index)
+            {
+              // If the current entry has already been deserialised, skip the
+              // payload for that entry
+              LOG_INFO_FMT(
+                "Skipping index {} as we are at index {}",
+                i,
+                append_entries_index);
+              ledger->skip_entry(data, size);
+              continue;
+            }
+
+            auto ret = ledger->get_entry(data, size);
+
+            if (!ret.second)
+            {
+              // NB: This will currently never be triggered.
+              // This should only fail if there is malformed data. Truncate
+              // the log and reply false.
+              LOG_FAIL_FMT(
+                "Recv append entries to {} from {} but the data is malformed",
+                local_id,
+                r.from_node);
+              ledger->truncate(r.prev_idx);
+              return;
+            }
+
+            ccf::Store::Tx tx;
+            auto deserialise_success = store->deserialise_views(
+              ret.first, public_only, false, nullptr, &tx);
+
+            switch (deserialise_success)
+            {
+              case kv::DeserialiseSuccess::FAILED:
+              {
+                LOG_FAIL_FMT("Replica failed to apply log entry {}", i);
+                break;
+              }
+              case kv::DeserialiseSuccess::PASS:
+              {
+                message_receiver_base->playback_transaction(tx);
+                break;
+              }
+              case kv::DeserialiseSuccess::PASS_SIGNATURE:
+              {
+                throw std::logic_error(
```

Comment:
I wonder if the kv should return a similar status when it has deserialised a `Pre_Prepare` transaction (in my mental model, PBFT's Pre-prepares are roughly Raft's signatures). Then you can remove `playback_transaction()` from `Replica.cpp` and call `playback_request()` and `playback_pre_prepare()` directly from here, which I think is neater?

Reply:
That would entail having the kv search look for `ccf.signature` table and `ccf.pbft.preprepares` table and return `PASS_SIGNATURE` status. I am weary of semantically mixing up pre-prepares with signatures. A pre-prepare is not necessarily signed and therefore it will not necessarily globally commit. The code will work, but the semantics will be off.

Reply:
We can return `PASS_PRE_PREPARE` from deserialise to keep the two enum different.

Reply:
Sure if we can agree on that I can add the `PASS_PRE_PREPARE` to `DeserialiseSuccess` and there will be no need for the playback_transaction as you said